pkg/mutation/mutation_function.go

```diff
@@ -0,0 +1,130 @@
+package mutation
+
+import (
+	"fmt"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/path/parser"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var log = logf.Log.WithName("mutation")
+
+func Mutate(mutator Mutator, obj *unstructured.Unstructured) error {
+	return mutate(mutator, obj.Object, nil, 0)
+}
+
+func mutate(m Mutator, current interface{}, previous interface{}, depth int) error {
+	if len(m.Path().Nodes)-1 == depth {
+		return addValue(m, current, previous, depth)
+	}
+	pathEntry := m.Path().Nodes[depth]
+	switch t := pathEntry.Type(); t {
+	case parser.ObjectNode:
+		next, ok := current.(map[string]interface{})[pathEntry.(*parser.Object).Reference]
+		if !ok {
+			next = createMissingElement(m, current, previous, depth)
+		}
+		if err := mutate(m, next, current, depth+1); err != nil {
+			return err
+		}
+		return nil
+	case parser.ListNode:
+		elementFound := false
+		listPathEntry := pathEntry.(*parser.List)
+		glob := listPathEntry.Glob
+		key := listPathEntry.KeyField
+		for _, listElement := range current.([]interface{}) {
+			if glob {
+				if err := mutate(m, listElement, current, depth+1); err != nil {
+					return err
+				}
+				elementFound = true
+			} else if elementValue, ok := listElement.(map[string]interface{})[key]; ok {
+				if *listPathEntry.KeyValue == elementValue {
+					if err := mutate(m, listElement, current, depth+1); err != nil {
+						return err
+					}
+					elementFound = true
+				}
+			}
+		}
+		// If no matching element in the array was found in non Globbed list, create a new element
+		if !listPathEntry.Glob && !elementFound {
+			next := createMissingElement(m, current, previous, depth)
+			if err := mutate(m, next, current, depth+1); err != nil {
+				return err
+			}
+		}
+	default:
+		return fmt.Errorf("Unrecognized type: %v", t)
+	}
+	return nil
+}
+
+func addValue(m Mutator, current interface{}, previous interface{}, depth int) error {
+	// TODO: it should be considered if the value set can be not just a simple string, but json which could be unmarshalled into an object
+	pathEntry := m.Path().Nodes[depth]
+	switch t := pathEntry.Type(); t {
+	case parser.ObjectNode:
+
+		switch m.(type) {
```

Comment:
That we're creating type switches here suggests to me that `addValue()` could be a property of the Mutator object. That would avoid the need for type-specific logic.

Not too big a concern currently, as we only have two types, but maybe worth leaving a comment about this so we revisit the issue if the code starts getting more complex?